akd/src/history_tree_node.rs

```diff
@@ -266,373 +250,209 @@ impl HistoryTreeNode {
+                    // self.update_node_hash::<_, H>(storage, epoch).await?;
+                    new_node.update_node_hash::<_, H>(storage, epoch).await?;
+                    parent.update_node_hash::<_, H>(storage, epoch).await?;
+                } else {
+                    // If no hashing, we need to manually save the nodes.
+                    new_leaf.write_to_storage(storage).await?;
+                    new_node.write_to_storage(storage).await?;
+                    parent.write_to_storage(storage).await?;
                 }
-                debug!("BEGIN save new_node");
-                new_node.write_to_storage(storage).await?;
-                debug!("BEGIN save parent");
-                parent.write_to_storage(storage).await?;
-                debug!("BEGIN retrieve new self");
-                *self =
-                    HistoryTreeNode::get_from_storage(storage, &NodeKey(self.label), epoch).await?;
                 debug!("END insert single leaf (dir_self = Some)");
                 Ok(())
             }
+            // Case where the current node is equal to the lcs
+            // Recurse!
             None => {
-                // case where the current node is equal to the lcs
-                debug!("BEGIN get child at epoch");
-                let child_st = self
-                    .get_child_at_epoch::<_, H>(storage, self.get_latest_epoch(), dir_leaf)
-                    .await?
-                    .ok_or_else(|| {
-                        HistoryTreeNodeError::NoChildAtEpoch(
-                            self.get_latest_epoch(),
-                            dir_leaf.unwrap_or(0),
-                        )
-                    })?;
-
                 debug!("BEGIN get child node from storage");
-                let mut child_node =
-                    HistoryTreeNode::get_from_storage(storage, &NodeKey(child_st.label), epoch)
-                        .await?;
+                let child_node = self.get_child_state(storage, dir_leaf).await?;
                 debug!("BEGIN insert single leaf helper");
-                child_node
-                    .insert_single_leaf_helper::<_, H>(storage, new_leaf, epoch, num_nodes, hashing)
-                    .await?;
-                if hashing {
-                    debug!("BEGIN update hashes");
-                    *self = HistoryTreeNode::get_from_storage(storage, &NodeKey(self.label), epoch)
-                        .await?;
-                    self.update_hash::<_, H>(storage, epoch).await?;
-                    self.write_to_storage(storage).await?;
-                } else {
-                    debug!("BEGIN retrieve self");
-                    *self = HistoryTreeNode::get_from_storage(storage, &NodeKey(self.label), epoch)
-                        .await?;
-                }
-                debug!("END insert single leaf (dir_self = None)");
-                Ok(())
+                let result = match child_node {
```

Comment:
nit: Remove the assignment of result, and just have the match statement end the fn.